### PR TITLE
[Channel] Add more validation

### PR DESCRIPTION
**Pull request by NoResponseMate**

### CODE REVIEW

**Review by Rafikooo — APPROVED** [review]

File: config/services/listeners.xml [hunk 1]
```diff
@@ -20,5 +20,12 @@
             <argument type="service" id="Sylius\PriceHistoryPlugin\Application\Logger\PriceChangeLoggerInterface" />
             <tag name="doctrine.event_listener" event="onFlush" lazy="true" />
         </service>
+
+        <service id="Sylius\PriceHistoryPlugin\Infrastructure\EventListener\BeforeDenormalizationChannelValidationListener">
+            <argument type="service" id="validator" />
+            <argument type="service" id="sylius.custom_factory.channel" />
+            <argument>%sylius_price_history.channel.validation_groups%</argument>
+            <tag name="kernel.event_listener" event="kernel.request" method="onKernelRequest" priority="3" />
```
[the remark this answers is not in the document]
Comment (NoResponseMate) [reply config/services/listeners.xml:28]: In this form, it needs to go before the APIP's `DeserializeListener` which has priority 2

[the remark this answers is not in the document]
Comment (Rafikooo) [reply config/services/listeners.xml:28]: Thanks for the explanation

File: src/Infrastructure/EventListener/BeforeDenormalizationChannelValidationListener.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    public function onKernelRequest(RequestEvent $event): void
+    {
+        $request = $event->getRequest();
+        $parameters = $request->attributes->all();
+
+        if (
+            !isset($parameters['_api_resource_class']) ||
+            !is_a($parameters['_api_resource_class'], ChannelInterface::class, true) ||
```
Comment (coldic3) [inline src/Infrastructure/EventListener/BeforeDenormalizationChannelValidationListener.php:39]:
```suggestion
            !is_a($parameters['_api_resource_class'] ?? null, ChannelInterface::class, true) ||
```
WDYT?

File: spec/Infrastructure/EventListener/BeforeDenormalizationChannelValidationListenerSpec.php [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        ;
+
+        $codeViolationsList->count()->willReturn(0);
+        $mainViolationsList->addAll(Argument::any())->shouldNotBeCalled();
+        $mainViolationsList->count()->willReturn(0);
+
+        $this->onKernelRequest($event);
+    }
+
+    function it_throw_violation_exception_when_at_least_one_property_is_not_valid(
```
Comment (coldic3) [inline spec/Infrastructure/EventListener/BeforeDenormalizationChannelValidationListenerSpec.php:152]:
```suggestion
    function it_throws_violation_exception_when_at_least_one_property_is_not_valid(
```

File: spec/Infrastructure/EventListener/BeforeDenormalizationChannelValidationListenerSpec.php [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+        $context->getViolations()->willReturn($violationsList);
+
+        $validator->validatePropertyValue(Argument::cetera())->shouldNotBeCalled();
+        $violationsList->count()->willReturn(0);
+        $violationsList->addAll(Argument::any())->shouldNotBeCalled();
+
+        $this->onKernelRequest($event);
+    }
+
+    function it_does_nothing_when_no_request_content_causes_no_violations(
```
Comment (coldic3) [inline spec/Infrastructure/EventListener/BeforeDenormalizationChannelValidationListenerSpec.php:116]:
```suggestion
    function it_does_not_throw_validation_exception_when_no_request_content_causes_no_violations(
```

File: spec/Infrastructure/EventListener/BeforeDenormalizationChannelValidationListenerSpec.php [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        $mainViolationsList->count()->willReturn(1);
+
+        // The 2 calls below are here to silence ValidatorException::__toString
+        $mainViolationsList->valid()->willReturn(false);
+        $mainViolationsList->rewind()->shouldBeCalled();
+
+        $this->shouldThrow(ConstraintViolationListAwareExceptionInterface::class)->during('onKernelRequest', [$event]);
+    }
+
+    function it_throw_violation_exception_when_all_properties_are_not_valid(
```
Comment (coldic3) [inline spec/Infrastructure/EventListener/BeforeDenormalizationChannelValidationListenerSpec.php:202]:
```suggestion
    function it_throws_violation_exception_when_all_properties_are_not_valid(
```

**Review by GSadee — APPROVED** [review]

**Pull request merged**